lib/src/st2110/experimental/st40_pipeline_rx.c

```diff
@@ -0,0 +1,707 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright(c) 2025 Intel Corporation
+ */
+
+#include "st40_pipeline_rx.h"
+
+#include "../../mt_log.h"
+#include "../../mt_stat.h"
+
+static const char* st40p_rx_frame_stat_name[ST40P_RX_FRAME_STATUS_MAX] = {
+    "free",
+    "ready",
+    "in_user",
+};
+
+static const char* rx_st40p_stat_name(enum st40p_rx_frame_status stat) {
+  return st40p_rx_frame_stat_name[stat];
+}
+
+static uint16_t rx_st40p_next_idx(struct st40p_rx_ctx* ctx, uint16_t idx) {
+  /* point to next */
+  uint16_t next_idx = idx;
+  next_idx++;
+  if (next_idx >= ctx->framebuff_cnt) next_idx = 0;
+  return next_idx;
+}
+
+static void rx_st40p_block_wake(struct st40p_rx_ctx* ctx) {
+  /* notify block */
+  mt_pthread_mutex_lock(&ctx->block_wake_mutex);
+  mt_pthread_cond_signal(&ctx->block_wake_cond);
+  mt_pthread_mutex_unlock(&ctx->block_wake_mutex);
+}
+
+static void rx_st40p_notify_frame_available(struct st40p_rx_ctx* ctx) {
+  if (ctx->ops.notify_frame_available) { /* notify app */
+    ctx->ops.notify_frame_available(ctx->ops.priv);
+  }
+
+  if (ctx->block_get) {
+    /* notify block */
+    rx_st40p_block_wake(ctx);
+  }
+}
+
+static struct st40p_rx_frame* rx_st40p_next_available(
+    struct st40p_rx_ctx* ctx, uint16_t idx_start, enum st40p_rx_frame_status desired) {
+  uint16_t idx = idx_start;
+  struct st40p_rx_frame* framebuff;
+
+  /* check ready frame from idx_start */
+  while (1) {
+    framebuff = &ctx->framebuffs[idx];
+    if (desired == framebuff->stat) {
+      /* find one desired */
+      return framebuff;
+    }
+    idx = rx_st40p_next_idx(ctx, idx);
+    if (idx == idx_start) {
+      /* loop all frames end */
+      break;
+    }
+  }
+
+  /* no any desired frame */
+  return NULL;
+}
+
+static int rx_st40p_rtp_ready(void* priv) {
+  struct st40p_rx_ctx* ctx = priv;
+  struct st40p_rx_frame* framebuff;
+  void* mbuf;
+  void* usrptr = NULL;
+  uint16_t len = 0;
+  struct st40_rfc8331_rtp_hdr* hdr;
+  struct st40_frame_info* frame_info;
+  uint32_t anc_count;
+  uint8_t* payload;
+  struct st40_rfc8331_payload_hdr* payload_hdr;
+  uint32_t payload_offset = 0;
+
+  if (!ctx->ready) return -EBUSY; /* not ready */
+
+  /* get mbuf from transport */
+  mbuf = st40_rx_get_mbuf(ctx->transport, &usrptr, &len);
+  if (!mbuf) return -EBUSY;
+
+  hdr = (struct st40_rfc8331_rtp_hdr*)usrptr;
+  anc_count = hdr->first_hdr_chunk.anc_count;
```

Comment:
please check if `len` is big enough to be able to access `hdr->first_hdr_chunk.anc_count`